### ISSUE TITLE
GDScript FileAccess reference scope bug

### ISSUE THREAD

**Issue by bitsawer** [godotengine/godot#73355]
### Godot version

4.0 dev d2b1474da79a4dce5c2031b3a3fafe8aaa2a161f

### System information

Windows 10

### Issue description

It seems like scoped reference objects like FileAccess sometimes live longer than they should, leading to all kind of strange behaviors. Mostly related to file handling encountering weird causality issues, like seeing temporary files that should no longer exist, files not existing even though you just wrote to them, reading back partially written files even though they should be already closed etc. Needless to say, these can cause subtle bugs and corrupted data in tool scripts. These are some reports that I suspect are caused by this bug:

* https://github.com/godotengine/godot/issues/73311
* https://github.com/godotengine/godot/issues/69551
* https://github.com/godotengine/godot/issues/68614
* https://github.com/godotengine/godot/issues/66818

This issue probably also applies to other classes than FileAccess, but it's the easiest place to see these things happen especially as the files are now closed by dropping all references to it instead of calling .close() which makes this bug easy to encounter.

In short, code like:

``` gdscript
var file = FileAccess.open("res://somefile.something", FileAccess.WRITE)
...
file = null
```

does not always free the reference and close the file, sometimes it happens only once the function/variable scope ends.

### Steps to reproduce

Open the script editor and run test_editor_script.gd (right click -> Run). Observe output. Uncomment test_2_ok(), run again and observe correct output. You can also run the project which calls the same code.

Make sure "filesystem/on_save/safe_save_on_backup_then_rename" is enabled (in Editor -> Editor Settings... -> FileSystem -> On Save). Then restart editor. Enabling temp files in not required, it just makes the bug easier to see in action. The test projects prints these temp files so the issue is easier to see as the temp files are renamed to overwrite the real file when the FileAccess object is actually closed. I guess this might affect a lot of scoped reference objects, but it's probably easiest to see when it manifests in external resources like files.

Optionally, apply this patch to the Godot source code C++ file access and rebuild the engine. It simply logs when a file with a certain type of name is actually closed. Note that this one is only for Windows FileAccess, each platform has its own.

```
@@ -155,10 +155,14 @@ Error FileAccessWindows::open_internal(const String &p_path, int p_mode_flags) {
 void FileAccessWindows::_close() {
 	if (!f) {
 		return;
 	}
 
+	if (path.find("bugfile") != -1) {
+		print_line("Close file: " + path);
+	}
+
 	fclose(f);
 	f = nullptr;
 
 	if (!save_path.is_empty()) {
 		bool rename_error = true;
```

Then run the test project.

``` gdscript
func test_fails():
    var file = FileAccess.open("res://bugfile_1.noextension", FileAccess.WRITE)
    var file2 = FileAccess.open("res://bugfile_2.noextension", FileAccess.WRITE)

    #These do not close both files, patch FileAccess C++ code to log file close calls.
    #Only first file is closed, file2 is closed only after function scope ends.
    file = null
    file2 = null

    print("The End")
```

With the patched engine this produces output

```
Close file: E:/Repositories/projects/file-close-bug/bugfile_1.noextension.tmp
The End
Close file: E:/Repositories/projects/file-close-bug/bugfile_2.noextension.tmp
```

Huh. Note that file2 is closed only after the print("The End") and once the function scope ends. Simply adding a junk reference counted variable "fixes" this:

``` gdscript
func test():
    var file = FileAccess.open("res://bugfile_1.noextension", FileAccess.WRITE)
    var file2 = FileAccess.open("res://bugfile_2.noextension", FileAccess.WRITE)

    #Simply adding this will close both files correctly and cleans up temp files
    #before function scope ends. Moving this before file creation will restore buggy file2 close behavior.
    var useless = RefCounted.new()

    file = null
    file2 = null

    print("The End")
```

This produces correct output:

```
Close file: E:/Repositories/projects/file-close-bug/bugfile_1.noextension.tmp
Close file: E:/Repositories/projects/file-close-bug/bugfile_2.noextension.tmp
The End
```

Looks like the last reference counted object in the stack has some lifetime issues or something similar? Using an int or string instead of RefCounted doesn't seem to work to "fix" the second test. It's kind of hard to say what is the root cause for this behavior. At first I was even thinking maybe print()'s were just buffered in wrong order somehow, but in the test project you can see the behavior with temp files, too.

### Minimal reproduction project

[file-close-bug.zip](https://github.com/godotengine/godot/files/10743162/file-close-bug.zip)

**Comment by bruvzg** [comment]
It's not a `FileAccess` specific issue, I can reproduce it with any Resource or RefCounted object in GDScript.

**Comment by bruvzg** [comment]
Done some debugging and seems like GDScript VM never clean up `stack`, so an extra reference will sit in the return value slot until it's overwritten by some other opcode. I'm not sure where it should be cleaned.

**Comment by dalexeev** [comment]
We should check this and the mentioned issues (#73311, #69551, #68614, #66818) since #73435 has been merged to separate issues caused by the error-prone `FileAccess` API from issues in GDScript.

**Comment by RandomShaper** [comment]
I've checked all of them with the fix that allows automatic closing of the files once out of scope. I'm certain that calling `close()` explicitly will work as fine.
